Build lapack failed since cmake/lapack.cmake is not new enough

Hello, I am trying to build OpenBLAS as a static C library. But I found that cmake/lapack.cmake is just not new enough compared with the build-in lapack-netlib

For example

<img width="1634" alt="Image" src="https://github.com/user-attachments/assets/e56144eb-e729-4977-a930-42fdf9cdefd9" />
The left side is from lapack-netlib/SRC/CMakeLists.txt  and right side is from cmake/lapack.cmake

The gap between the script makes building `libopenblas.a` failed since there are some undefined symbols in lapack.

@martin-frbg Thanks for the reply. 

> are you building 0.3.29 or the develop branch?

I am using commit `200771078f337d6b20ed11315d2d60c075bb6263` on branch develop

> More detail

I am building a faiss demo by using OpenBLAS

```text
set(BUILD_WITHOUT_LAPACKE OFF)
set(NO_SHARED 1)
set(NO_AVX512 1)
set(NO_LAPACK FALSE)
set(C_LAPACK ON)
set(USE_OPENMP TRUE)
set(BUILD_STATIC_LIBS ON)
set(NOFORTRAN TRUE)
set(BUILD_TESTING OFF)
set(USE_OPTIMIZED_LAPACK OFF)
add_subdirectory(contrib/OpenBLAS)

set(FAISS_ENABLE_GPU OFF)
set(FAISS_ENABLE_PYTHON OFF)

add_subdirectory(contrib/faiss)
add_executable(vector-search main.cpp)
target_link_libraries(vector-search faiss)
```
And the output is
```text
[ 97%] Linking C executable openblas_utest
[ 98%] Built target faiss
[ 98%] Linking C executable openblas_utest_ext
[ 98%] Linking CXX executable vector-search
ld.lld: error: undefined symbol: cpotrf_
>>> referenced by test_potrs.c
>>>               CMakeFiles/openblas_utest.dir/test_potrs.c.o:(__ctest_potrf_bug_695_run)
>>> referenced by test_potrs.c
>>>               CMakeFiles/openblas_utest.dir/test_potrs.c.o:(__ctest_potrf_bug_695_run)
>>> referenced by test_potrs.c
>>>               CMakeFiles/openblas_utest.dir/test_potrs.c.o:(__ctest_potrf_smoketest_trivial_run)

ld.lld: error: undefined symbol: spotrf_
>>> referenced by test_potrs.c
>>>               CMakeFiles/openblas_utest.dir/test_potrs.c.o:(__ctest_potrf_smoketest_trivial_run)

ld.lld: error: undefined symbol: dpotrf_
>>> referenced by test_potrs.c
>>>               CMakeFiles/openblas_utest.dir/test_potrs.c.o:(__ctest_potrf_smoketest_trivial_run)

ld.lld: error: undefined symbol: zpotrf_
>>> referenced by test_potrs.c
>>>               CMakeFiles/openblas_utest.dir/test_potrs.c.o:(__ctest_potrf_smoketest_trivial_run)
collect2: error: ld returned 1 exit status
make[2]: *** [contrib/OpenBLAS/utest/CMakeFiles/openblas_utest.dir/build.make:360: contrib/OpenBLAS/utest/openblas_utest] Error 1
make[1]: *** [CMakeFiles/Makefile2:831: contrib/OpenBLAS/utest/CMakeFiles/openblas_utest.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
ld.lld: error: undefined symbol: zspmv_
>>> referenced by test_zspmv.c
>>>               CMakeFiles/openblas_utest_ext.dir/test_extensions/test_zspmv.c.o:(check_zspmv)
>>> referenced by test_zspmv.c
>>>               CMakeFiles/openblas_utest_ext.dir/test_extensions/test_zspmv.c.o:(check_badargs)
>>> did you mean: zsbmv_
>>> defined in: ../lib/libopenblas.a(zsbmv.c.o)

ld.lld: error: undefined symbol: cspmv_
>>> referenced by test_cspmv.c
>>>               CMakeFiles/openblas_utest_ext.dir/test_extensions/test_cspmv.c.o:(check_cspmv)
>>> referenced by test_cspmv.c
>>>               CMakeFiles/openblas_utest_ext.dir/test_extensions/test_cspmv.c.o:(check_badargs)
collect2: error: ld returned 1 exit status
make[2]: *** [contrib/OpenBLAS/utest/CMakeFiles/openblas_utest_ext.dir/build.make:1096: contrib/OpenBLAS/utest/openblas_utest_ext] Error 1
make[1]: *** [CMakeFiles/Makefile2:863: contrib/OpenBLAS/utest/CMakeFiles/openblas_utest_ext.dir/all] Error 2
ld.lld: error: undefined symbol: sgetrf_
>>> referenced by LocalSearchQuantizer.cpp
>>>               LocalSearchQuantizer.cpp.o:((anonymous namespace)::fmat_inverse(float*, int)) in archive contrib/faiss/faiss/libfaiss.a
>>> did you mean: sgeqrf_
>>> defined in: contrib/OpenBLAS/lib/libopenblas.a(sgeqrf.c.o)

ld.lld: error: undefined symbol: dgetrf_
>>> referenced by LocalSearchQuantizer.cpp
>>>               LocalSearchQuantizer.cpp.o:((anonymous namespace)::dmat_inverse(double*, int)) in archive contrib/faiss/faiss/libfaiss.a
>>> did you mean: dgeqrf_
>>> defined in: contrib/OpenBLAS/lib/libopenblas.a(dgeqrf.c.o)

ld.lld: error: undefined symbol: strtri_
>>> referenced by sgetri.c
>>>               sgetri.c.o:(sgetri_) in archive contrib/OpenBLAS/lib/libopenblas.a

ld.lld: error: undefined symbol: dtrtri_
>>> referenced by dgetri.c
>>>               dgetri.c.o:(dgetri_) in archive contrib/OpenBLAS/lib/libopenblas.a
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/vector-search.dir/build.make:105: vector-search] Error 1
make[1]: *** [CMakeFiles/Makefile2:471: CMakeFiles/vector-search.dir/all] Error 2
```
I tried to fix the undefiend symbols one by one until there is not error, and the final fix is as https://github.com/OpenMathLib/OpenBLAS/pull/5194

I tried to set `BUILD_RELAPACK` to true, but the problem still exists.

<img width="577" alt="Image" src="https://github.com/user-attachments/assets/8e006ea7-c579-4603-b02d-e92bfd0d90d5" />

```
ld.lld: error: undefined symbol: sgetrf_
>>> referenced by LocalSearchQuantizer.cpp
>>>               LocalSearchQuantizer.cpp.o:((anonymous namespace)::fmat_inverse(float*, int)) in archive contrib/faiss/faiss/libfaiss.a
>>> did you mean: sgeqrf_
>>> defined in: contrib/OpenBLAS/lib/libopenblas.a(sgeqrf.c.o)

ld.lld: error: undefined symbol: dgetrf_
>>> referenced by LocalSearchQuantizer.cpp
>>>               LocalSearchQuantizer.cpp.o:((anonymous namespace)::dmat_inverse(double*, int)) in archive contrib/faiss/faiss/libfaiss.a
>>> did you mean: dgeqrf_
>>> defined in: contrib/OpenBLAS/lib/libopenblas.a(dgeqrf.c.o)

ld.lld: error: undefined symbol: strtri_
>>> referenced by sgetri.c
>>>               sgetri.c.o:(sgetri_) in archive contrib/OpenBLAS/lib/libopenblas.a

ld.lld: error: undefined symbol: dtrtri_
>>> referenced by dgetri.c
>>>               dgetri.c.o:(dgetri_) in archive contrib/OpenBLAS/lib/libopenblas.a
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/vector-search.dir/build.make:105: vector-search] Error 1
make[1]: *** [CMakeFiles/Makefile2:491: CMakeFiles/vector-search.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[100%] Built target d1
```
I checked the archive file
```
File: libopenblas.a(sgetrf.c.o)

Symbol table '.symtab' contains 18 entries:
   Num:    Value          Size Type    Bind   Vis      Ndx Name
     0: 0000000000000000     0 NOTYPE  LOCAL  DEFAULT  UND 
     1: 0000000000000000     0 FILE    LOCAL  DEFAULT  ABS sgetrf.c
     2: 0000000000000000     0 SECTION LOCAL  DEFAULT    1 
     3: 0000000000000000     0 SECTION LOCAL  DEFAULT    3 
     4: 0000000000000000     0 SECTION LOCAL  DEFAULT    4 
     5: 0000000000000000     0 SECTION LOCAL  DEFAULT    5 
     6: 0000000000000215  1002 FUNC    LOCAL  DEFAULT    1 RELAPACK_sgetrf_rec
     7: 0000000000000000     0 SECTION LOCAL  DEFAULT    7 
     8: 0000000000000000     0 SECTION LOCAL  DEFAULT    8 
     9: 0000000000000000     0 SECTION LOCAL  DEFAULT    6 
    10: 0000000000000000   533 FUNC    GLOBAL DEFAULT    1 RELAPACK_sgetrf
    11: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND _GLOBAL_OFFSET_TABLE_
    12: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND xerbla_
    13: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND slaswp_
    14: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND strsm_
    15: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND __stack_chk_fail
    16: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND sgetrf2_
    17: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND sgemm_
```
I do have an object file sgetrf.c.o, it seems this is compiled from `relapack/src/strtri.c` which has an impl of `RELAPACK_sgetrf` instead of `sgetrf_`.
```
void RELAPACK_strtri(
    const char *uplo, const char *diag, const blasint *n,
    float *A, const blasint *ldA,
    blasint *info
) {
...
}
```
Is there any scripts to convert the function name in relapack to standard lapack function?
@martin-frbg

--- a/interface/CMakeLists.txt
+++ b/interface/CMakeLists.txt
@@ -30,17 +30,17 @@ set(BLAS2_SOURCES
   gemv.c ger.c
   trsv.c trmv.c 
   syr2.c gbmv.c
-  sbmv.c 
+  sbmv.c spmv.c
   spr2.c
   tbsv.c tbmv.c
   tpsv.c tpmv.c
 )
 
 set(BLAS2_REAL_ONLY_SOURCES
-  symv.c syr.c spmv.c spr.c
+  symv.c syr.c spr.c
 )
 set(BLAS2_COMPLEX_LAPACK_SOURCES
-  symv.c syr.c spmv.c spr.c
+  symv.c syr.c spr.c
 )
 
 set(BLAS2_COMPLEX_ONLY_MANGLED_SOURCES
@@ -195,7 +195,7 @@ if (NOT DEFINED NO_CBLAS)
   endforeach ()
 endif()
 
-if (NOT DEFINED NO_LAPACK)
+if (NOT NO_LAPACK)
   set(LAPACK_SOURCES
     lapack/gesv.c
   )

